File: gpa.py
"""
南理工课表管理系统 — 绩点计算模块
==============================
成绩→绩点换算（NJUST 4.0 量表）和加权平均 GPA 计算。
"""
from collections import OrderedDict

# 非正式成绩状态（不参与绩点计算）
_NON_GRADE_STATUS = {"缓考", "缺考", "免修", "作弊", "违纪", "取消", "旷考", "休学"}

# 不计入 GPA 的课程性质（NJUST 教务统一口径：通识公共选修课不参与评奖/保研绩点）
_NON_GPA_NATURES = {"通识教育选修课"}


def is_gpa_course(course_nature: str) -> bool:
    """判断课程是否计入平均学分绩点（评奖、保研、学籍审核）

    不计入：通识教育选修课（校公选、素质拓展类）
    计入：必修课、专业选修课、专业限选课等所有其他课程
    """
    return (course_nature or "").strip() not in _NON_GPA_NATURES


def score_to_gp(score) -> float:
    """将百分制成绩或等级制成绩转换为绩点（NJUST 4.0 量表）

    返回值：
    - 正数 / 0：有效绩点（0 = 不及格，仍计入加权平均）
    - -1：非正式成绩（缓考/缺考/免修等，不参与计算）
    """
    # 文字等级（NJUST 五级制 + 带加减的扩展等级）
    level_map = {
        # 优秀档（90-100）
        "优": 4.0, "优秀": 4.0,
        "优-": 3.7, "优秀-": 3.7,
        # 良好档（80-89）
        "良+": 3.3, "良好+": 3.3,
        "良": 3.0, "良好": 3.0,
        "良-": 2.7, "良好-": 2.7,
        # 中等档（70-79）
        "中+": 2.3, "中等+": 2.3,
        "中": 2.0, "中等": 2.0,
        "中-": 1.5, "中等-": 1.5,
        # 及格档（60-69）
        "及格": 1.0, "通过": 1.0,
        # 不及格
        "不及格": 0, "不通过": 0,
    }
    s = str(score).strip()
    if s in level_map:
        return level_map[s]
    # 非正式成绩状态 → 排除
    if s in _NON_GRADE_STATUS:
        return -1.0
    # 百分制 → 绩点（NJUST 官方换算表，.0/.5 精度）
    try:
        v = float(s)
    except (ValueError, TypeError):
        return -1.0  # 无法识别的文字（如教务系统特殊标记），排除
    if v >= 90:
        return 4.0   # 90.0~100.0 → 优
    if v >= 85:
        return 3.7   # 85.0~89.5  → 优-
    if v >= 82:
        return 3.3   # 82.0~84.5  → 良+
    if v >= 78:
        return 3.0   # 78.0~81.5  → 良
    if v >= 75:
        return 2.7   # 75.0~77.5  → 良-
    if v >= 72:
        return 2.3   # 72.0~74.5  → 中+
    if v >= 68:
        return 2.0   # 68.0~71.5  → 中
    if v >= 64:
        return 1.5   # 64.0~67.5  → 中-
    if v >= 60:
        return 1.0   # 60.0~63.5  → 及格
    return 0                   # < 60.0     → 不及格


def calc_gpa(grades: list[dict], gpa_only: bool = True) -> float:
    """计算加权平均绩点 Σ(学分×绩点) / Σ学分

    不及格课程（绩点=0）计入分母，拉低平均；
    缓考/缺考/免修等非正式成绩（绩点=-1）不计入。

    参数:
        grades: 成绩列表，每项含 credit, grade_point, score
        gpa_only: True 时排除通识教育选修课（评奖/保研口径）
    """
    total_weighted = 0.0
    total_credits = 0.0
    for g in grades:
        if gpa_only and not is_gpa_course(g.get("course_nature", "")):
            continue
        credit = float(g.get("credit", 0) or 0)
        if credit <= 0:
            continue
        gp = float(g.get("grade_point", 0) or 0)
        if gp == 0:
            # 数据库无绩点，从成绩换算（可能返回 -1 表示排除）
            gp = score_to_gp(g.get("score", ""))
        if gp >= 0:
            total_weighted += credit * gp
            total_credits += credit
    return round(total_weighted / total_credits, 2) if total_credits > 0 else 0.0


def calc_semester_gpas(grades: list[dict]) -> list[dict]:
    """计算每个学期的绩点汇总

    参数:
        grades: 成绩字典列表，每项含 academic_year, semester, course_name,
                score, credit, grade_point, course_nature

    返回按学期排序的列表: [{semester, gpa, gpa_all, credits, count}, ...]
    其中 gpa 为排除通识选修课后的计内绩点（评奖/保研口径）
    """
    # 按学期分组
    groups = OrderedDict()
    for r in grades:
        key = f"{r.get('academic_year', '')}-{r.get('semester', '')}"
        if key not in groups:
            groups[key] = []
        groups[key].append({
            "course_name": r.get("course_name", ""),
            "score": r.get("score", ""),
            "credit": r.get("credit", 0),
            "grade_point": r.get("grade_point", 0),
            "course_nature": r.get("course_nature", "") or "",
        })

    result = []
    for sem, items in groups.items():
        # 补充绩点
        for g in items:
            if float(g.get("grade_point", 0) or 0) == 0:
                g["grade_point"] = score_to_gp(g.get("score", ""))

        gpa_all = calc_gpa(items, gpa_only=False)
        gpa_counted = calc_gpa(items, gpa_only=True)

        counted_items = [g for g in items if is_gpa_course(g.get("course_nature", ""))]
        counted_credits = round(sum(
            float(g.get("credit", 0) or 0) for g in counted_items
        ), 1)

        result.append({
            "semester": sem,
            "gpa": gpa_counted,
            "gpa_all": gpa_all,
            "credits": counted_credits,
            "count": len(counted_items),
        })

    return result


# ================================================================
# 四六级折算 & 保研 GPA 模式
# ================================================================

def cet_to_percentage(cet_score: float, cet_type: str) -> float:
    """四六级分数 → 百分制折算（NJUST 官方公式）

    CET4: 基础分 = (分数 - 425) / 285 × 40 + 60
    CET6: 基础分 + 5（封顶 100）

    返回: 百分制分数（60~100），< 425 返回 0 表示不可用
    """
    if cet_score < 425:
        return 0.0
    base = (cet_score - 425) / 285.0 * 40 + 60
    if cet_type == "CET6":
        base = min(base + 5, 100.0)
    return round(base, 1)


def is_english_course(course_name: str) -> bool:
    """判断是否为英语课（保研模式中被 CET 替换）

    匹配: "通用英语" 或 "专用英语-*"
    """
    name = course_name.strip()
    return name == "通用英语" or name.startswith("专用英语-")


def calc_gpa_baoyan(grades: list[dict], cet_scores: list[dict] = None,
                     gpa_only: bool = True) -> float:
    """保研/推免模式绩点：用四六级折算分替换英语模块（8学分）

    规则:
    1. 从 grades 中移除所有英语课（通用英语 + 专用英语-*）
    2. 取最高可用 CET 分折算成百分制，按 8 学分插入
    3. 优先级: CET6 > CET4，都不可用则回退到原始英语课成绩

    参数:
        grades: 成绩列表
        cet_scores: 四六级成绩 [{type, score}, ...]，None 表示不替换
        gpa_only: 是否排除通识选修课
    """
    if not grades:
        return 0.0

    # 分离英语课和非英语课
    english_grades = []
    non_english = []
    for g in grades:
        if is_english_course(g.get("course_name", "")):
            english_grades.append(g)
        else:
            non_english.append(g)

    # 选择最佳 CET 分数
    cet4_score = 0.0
    cet6_score = 0.0
    if cet_scores:
        for cs in cet_scores:
            if cs.get("type") == "CET4":
                cet4_score = max(cet4_score, float(cs.get("score", 0) or 0))
            elif cs.get("type") == "CET6":
                cet6_score = max(cet6_score, float(cs.get("score", 0) or 0))

    # 优先 CET6，其次 CET4
    pct = 0.0
    if cet6_score >= 425:
        pct = cet_to_percentage(cet6_score, "CET6")
    elif cet4_score >= 425:
        pct = cet_to_percentage(cet4_score, "CET4")

    if pct > 0:
        # 有可用 CET 分数 → 用 CET 折算分替换英语模块
        gp = score_to_gp(pct)
        cet_entry = {
            "course_name": "CET折算(英语模块)",
            "score": str(pct),
            "credit": 8.0,
            "grade_point": gp,
            "course_nature": "CET替换",
        }
        calc_grades = non_english + [cet_entry]
    else:
        # 无可用 CET 分数 → 回退原始英语课成绩
        calc_grades = list(grades)

    return calc_gpa(calc_grades, gpa_only=gpa_only)

File: utils/gpa.js
/**
 * GPA 计算工具 — 从服务端 gpa.py 移植到前端（方案 A）
 * NJUST 4.0 量表、学期/总 GPA、CET 折算、保研模式
 */

// 非正式成绩状态（不参与绩点计算）
const NON_GRADE_STATUS = ['缓考', '缺考', '免修', '作弊', '违纪', '取消', '旷考', '休学']

// 不计入 GPA 的课程性质（NJUST 教务统一口径：通识公共选修课不参与评奖/保研绩点）
const NON_GPA_NATURES = ['通识教育选修课']

// 文字等级（NJUST 五级制 + 带加减的扩展等级）
const LEVEL_MAP = {
  '优': 4.0, '优秀': 4.0, '优+': 4.0, '优秀+': 4.0,
  '优-': 3.7, '优秀-': 3.7,
  '良+': 3.3, '良好+': 3.3,
  '良': 3.0, '良好': 3.0,
  '良-': 2.7, '良好-': 2.7,
  '中+': 2.3, '中等+': 2.3,
  '中': 2.0, '中等': 2.0,
  '中-': 1.5, '中等-': 1.5,
  '及格': 1.0, '通过': 1.0,
  '不及格': 0, '不通过': 0
}

// 等级制 → 百分制折算(与参考 App 一致, 用于平均分):
// 优=90 优-=87 良+=83 良=80 良-=76 中+=73 中=70 中-=66 及格=60
const LEVEL_PCT = {
  '优': 90, '优秀': 90, '优+': 90, '优秀+': 90,
  '优-': 87, '优秀-': 87,
  '良+': 83, '良好+': 83,
  '良': 80, '良好': 80,
  '良-': 76, '良好-': 76,
  '中+': 73, '中等+': 73,
  '中': 70, '中等': 70,
  '中-': 66, '中等-': 66,
  '及格': 60, '通过': 60,
  '不及格': 0, '不通过': 0
}

/** 判断课程是否计入平均学分绩点（评奖、保研、学籍审核） */
function isGpaCourse(courseNature) {
  return !NON_GPA_NATURES.includes((courseNature || '').trim())
}

/** 百分制/等级制成绩 → 绩点（NJUST 4.0 量表）；非正式成绩返回 -1 */
function scoreToGp(score) {
  const s = String(score == null ? '' : score).trim()
  if (s in LEVEL_MAP) return LEVEL_MAP[s]
  if (NON_GRADE_STATUS.includes(s)) return -1
  const v = parseFloat(s)
  if (isNaN(v)) return -1
  if (v >= 90) return 4.0
  if (v >= 85) return 3.7
  if (v >= 82) return 3.3
  if (v >= 78) return 3.0
  if (v >= 75) return 2.7
  if (v >= 72) return 2.3
  if (v >= 68) return 2.0
  if (v >= 64) return 1.5
  if (v >= 60) return 1.0
  return 0
}

function _round(v, digits) {
  const n = parseFloat(v)
  return isNaN(n) ? 0 : Math.round(n * Math.pow(10, digits || 2)) / Math.pow(10, digits || 2)
}

/** 等级制/百分制成绩 → 百分制数值(等级按参考 App 折算表); 非正式成绩返回 null */
function scoreToPct(score) {
  const s = String(score == null ? '' : score).trim()
  if (s in LEVEL_PCT) return LEVEL_PCT[s]
  if (NON_GRADE_STATUS.includes(s)) return null
  const v = parseFloat(s)
  return isNaN(v) ? null : v
}

/**
 * 平均分 = Σ(百分制成绩 × 学分) / Σ学分(学分加权, 等级制折算参与)
 * 与参考 App 口径一致; 无有效成绩时返回 0
 */
function calcAvg(grades) {
  let weighted = 0
  let credits = 0
  for (const g of grades || []) {
    const pct = scoreToPct(g.score)
    if (pct == null) continue
    const credit = parseFloat(g.credit) || 0
    const wc = credit > 0 ? credit : 1   // 无学分课程按 1 计(与参考 App 一致)
    weighted += pct * wc
    if (credit > 0) credits += credit
  }
  if (credits > 0) return _round(weighted / credits)
  return weighted > 0 ? _round(weighted / (grades || []).length) : 0
}

/**
 * 加权平均绩点 Σ(学分×绩点) / Σ学分
 * 不及格（绩点=0）计入分母；缓考/缺考/免修（绩点=-1）不计入
 * @param {Array} grades 成绩列表，每项含 credit, grade_point, score, course_nature
 * @param {boolean} gpaOnly true 时排除通识教育选修课（评奖/保研口径）
 */
function calcGpa(grades, gpaOnly) {
  let totalWeighted = 0
  let totalCredits = 0
  for (const g of grades || []) {
    if (gpaOnly && !isGpaCourse(g.course_nature)) continue
    const credit = parseFloat(g.credit) || 0
    if (credit <= 0) continue
    let gp = parseFloat(g.grade_point) || 0
    if (gp === 0) gp = scoreToGp(g.score)
    if (gp >= 0) {
      totalWeighted += credit * gp
      totalCredits += credit
    }
  }
  return totalCredits > 0 ? _round(totalWeighted / totalCredits) : 0
}

/** 四六级总分 → 百分制折算(参考 App 公式: 总分/710×100, 下限 60); 无效分数返回 0 */
function cetToPercentage(cetScore, cetType) {
  const score = parseFloat(cetScore) || 0
  if (score <= 0) return 0
  const pct = score / 710 * 100
  return _round(Math.min(100, Math.max(60, pct)), 1)
}

/** 判断是否为英语课（保研模式中被 CET 替换） */
function isEnglishCourse(courseName) {
  const name = (courseName || '').trim()
  return name === '通用英语' || name.indexOf('专用英语-') === 0
}

/**
 * 保研/推免模式绩点: 四六级折算成绩以 8 学分附加计入(参考 App 口径,
 * 不替换英语课; 原英语课成绩保留参与计算)
 * 优先级: CET6 > CET4
 */
function calcGpaBaoyan(grades, cetScores, gpaOnly) {
  const list = grades || []
  if (list.length === 0) return 0

  let cet4 = 0
  let cet6 = 0
  for (const cs of (cetScores || [])) {
    const score = parseFloat(cs.score) || 0
    if (cs.type === 'CET4') cet4 = Math.max(cet4, score)
    else if (cs.type === 'CET6') cet6 = Math.max(cet6, score)
  }

  let pct = 0
  if (cet6 >= 425) pct = cetToPercentage(cet6, 'CET6')
  else if (cet4 >= 425) pct = cetToPercentage(cet4, 'CET4')

  let calcGrades = list
  if (pct > 0) {
    calcGrades = list.concat([{
      course_name: 'CET折算(8学分)',
      score: String(pct),
      credit: 8,
      grade_point: scoreToGp(pct),
      course_nature: 'CET'
    }])
  }
  return calcGpa(calcGrades, gpaOnly === undefined ? true : gpaOnly)
}

module.exports = {
  isGpaCourse,
  scoreToGp,
  scoreToPct,
  calcAvg,
  calcGpa,
  cetToPercentage,
  isEnglishCourse,
  calcGpaBaoyan
}
